File: 1381-design-a-stack-with-increment-operation/1381-design-a-stack-with-increment-operation.c
#include <stdlib.h>

/**
 * @brief Struct representing a custom stack.
 * 
 * @param stack (int*): Dynamic array to store stack elements.
 * @param stack_size (int): Current number of elements in the stack.
 * @param stack_capacity (int): Maximum capacity of the stack.
 */
typedef struct {
    int* stack;
    int stack_size;
    int stack_capacity;
} CustomStack;

/**
 * @brief Creates a new CustomStack with the specified maximum size.
 * 
 * @param maxSize (int): The maximum size (capacity) of the stack.
 * 
 * @return CustomStack*: Pointer to the newly created stack.
 */
CustomStack* customStackCreate(int maxSize) {
    CustomStack* mystack = (CustomStack*)malloc(sizeof(CustomStack));
    if (mystack == NULL) {
        return NULL;  // Handle memory allocation failure.
    }
    mystack->stack = (int*)malloc(sizeof(int) * maxSize);
    if (mystack->stack == NULL) {
        free(mystack);  // Free memory in case stack allocation fails.
        return NULL;
    }
    mystack->stack_size = 0;
    mystack->stack_capacity = maxSize;
    return mystack;
}

/**
 * @brief Push an element onto the stack if there is space.
 * 
 * @param obj (CustomStack*): Pointer to the CustomStack.
 * @param x (int): The element to be pushed onto the stack.
 */
void customStackPush(CustomStack* obj, int x) {
    if (obj != NULL && obj->stack_size < obj->stack_capacity) {
        obj->stack[obj->stack_size] = x;  // Add element at the current top
        obj->stack_size++;  // Increment size after adding the element
    }
}

/**
 * @brief Pop an element from the stack.
 * 
 * @param obj (CustomStack*): Pointer to the CustomStack.
 * 
 * @return int: The popped element, or -1 if the stack is empty.
 */
int customStackPop(CustomStack* obj) {
    if (obj != NULL && obj->stack_size != 0) {
        obj->stack_size--;  // Decrement size first
        return obj->stack[obj->stack_size];  // Return the element at the new top
    }
    return -1;  // Return -1 if the stack is empty.
}

/**
 * @brief Increment the bottom k elements of the stack by a given value.
 * 
 * @param obj (CustomStack*): Pointer to the CustomStack.
 * @param k (int): Number of elements from the bottom to be incremented.
 * @param val (int): The value to increment each of the bottom k elements by.
 */
void customStackIncrement(CustomStack* obj, int k, int val) {
    if (obj != NULL) {
        int limit = (k < obj->stack_size) ? k : obj->stack_size;  // Determine the limit of elements to increment.
        for (int i = 0; i < limit; i++) {
            obj->stack[i] += val;  // Increment each element.
        }
    }
}

/**
 * @brief Frees the memory allocated for the CustomStack.
 * 
 * @param obj (CustomStack*): Pointer to the CustomStack.
 */
void customStackFree(CustomStack* obj) {
    if (obj != NULL) {
        free(obj->stack);  // Free the memory allocated for the stack array.
        free(obj);  // Free the memory allocated for the CustomStack struct.
    }
}


/**
 * Your CustomStack struct will be instantiated and called as such:
 * CustomStack* obj = customStackCreate(maxSize);
 * customStackPush(obj, x);
 
 * int param_2 = customStackPop(obj);
 
 * customStackIncrement(obj, k, val);
 
 * customStackFree(obj);
*/

File: Difficulty/B. Medium/1381-design-a-stack-with-increment-operation/1381-design-a-stack-with-increment-operation.py
class CustomStack:
    """
    Custom stack with fixed size that supports push, pop, and increment operations.

    Attributes:
    stack (list): Internal list to store the elements of the stack.
    stack_size (int): Current size of the stack.
    stack_capacity (int): Maximum capacity of the stack.
    """

    def __init__(self, maxSize: int):
        """
        Initialize the stack with a given maximum size.

        Parameters:
        maxSize (int): The maximum size of the stack.
        """
        self.stack: list[int] = []
        self.stack_size: int = 0
        self.stack_capacity: int = maxSize

    def push(self, x: int) -> None:
        """
        Push an element onto the stack if it's not full.

        Parameters:
        x (int): The element to be pushed onto the stack.
        """
        if self.stack_size < self.stack_capacity:
            self.stack.append(x)
            self.stack_size += 1

    def pop(self) -> int:
        """
        Pop an element from the stack and return it. 
        If the stack is empty, return -1.

        Returns:
        int: The popped element if stack is not empty, otherwise -1.
        """
        if self.stack_size != 0:
            self.stack_size -= 1
            return self.stack.pop()
        else:
            return -1

    def increment(self, k: int, val: int) -> None:
        """
        Increment the bottom k elements of the stack by val. 
        If the stack has fewer than k elements, increment all of them.

        Parameters:
        k (int): The number of elements from the bottom to be incremented.
        val (int): The value to increment by.
        """
        # Increment up to min(k, stack_size) to handle case where stack has fewer than k elements
        limit = min(k, self.stack_size)
        for i in range(limit):
            self.stack[i] += val


# Your CustomStack object will be instantiated and called as such:
# obj = CustomStack(maxSize)
# obj.push(x)
# param_2 = obj.pop()
# obj.increment(k,val)